File: c/HandlerChain_MessageWindow.c
/// @file
/// @brief
/// Declaration of the @ref MessageWindow structure, along with the
/// implementation of the support functions, MessageWindow_Create(),
/// MessageWindow_Destroy(), MessageWindow_ProcessMessage(), and
/// MessageWindow_ToString(),
/// as used in the @ref handlerchain_pattern.


#include <stdlib.h>
#include <stdio.h>

#include "helpers/formatstring.h"

#include "HandlerChain_WindowRectangle.h"
#include "HandlerChain_HandlerFunctions.h"
#include "HandlerChain_MessageWindow.h"


const int CLOSE_WIDTH = 2; ///< Width of the close region in the window.
const int CLOSE_HEIGHT = 2; ///< Height of the close region in the window.

/// <summary>
/// Represents a window in an arbitrary space. It has an ID, title, and
/// position.  A close box is within the window's rectangle.  The window can be
///  selected by sending a button down message with the position within this
/// window.  If a button down message occurs outside this window, it is
/// deselected.
/// </summary>
struct _Window
{
    // Note: We must specify struct _Window so we can add pointer references for
    // next and prev in the structure.  These cannot be specified with the
    // typedef alias of MessageWindow.

    /// <summary>
    /// Previous window in a linked list.  NULL if this window is the head of the list.
    /// </summary>
    struct _Window* prev;

    /// <summary>
    /// Next window in a linked list.  NULL if this windows is the last in the list.
    /// </summary>
    struct _Window* next;

    /// <summary>
    /// Unique ID of this window.
    /// </summary>
    int _windowId;

    /// <summary>
    /// Title/Name of this window.
    /// </summary>
    const char* _title;

    /// <summary>
    /// Position of this window in global coordinates.
    /// </summary>
    WindowRectangle _windowBox;

    /// <summary>
    /// Position of the close window within the window box, although the
    /// coordinates are also global coordinates to eliminate the need to
    /// convert between window and global coordinates.
    /// </summary>
    WindowRectangle _closeBox;

    /// <summary>
    /// Whether this window has been selected (a button click occurred
    /// within the window).
    /// </summary>
    bool _selected;
};

/// <summary>
/// Alias for struct _Window, which represents a window in an arbitrary space.
/// It has an ID, title, and position.  A close box is within the window's
/// rectangle.  The window can be selected by sending a button down message
/// with the position within this window.  If a button down message occurs
/// outside this window, it is deselected.
/// </summary>
typedef struct _Window MessageWindow;


/// <summary>
/// List of all created @ref MessageWindow objects, in a double-linked list.
/// </summary>
static MessageWindow* _windowList = NULL;

/// <summary>
/// The next ID to use for a new window.
/// </summary>
static int _nextWindowId = 1;

/// <summary>
/// Helper function to append the given @ref MessageWindow to the list of
/// @ref MessageWindow objects.
/// </summary>
/// <param name="window">A @ref MessageWindow object to append to the list.</param>
static void _AppendWindowToList(MessageWindow* window)
{
    if (_windowList == NULL)
    {
        _windowList = window;
    }
    else
    {
        MessageWindow* nextWindow = _windowList;
        while (nextWindow->next != NULL)
        {
            nextWindow = nextWindow->next;
        }
        nextWindow->next = window;
        window->prev = nextWindow;
    }
}

/// <summary>
/// Helper function to remove the given @ref MessageWindow object from the
/// linked list of @ref MessageWindow objects.
/// </summary>
/// <param name="window">The @ref MessageWindow object to remove from the list.</param>
static void _RemoveWindowFromList(MessageWindow* window)
{
    if (window != NULL)
    {
        if (window->prev != NULL)
        {
            window->prev->next = window->next;

        }
        else
        {
            _windowList = window->next;
        }

        if (window->next != NULL)
        {
            window->next->prev = window->prev;
        }

        window->next = NULL;
        window->prev = NULL;
    }
}

/// <summary>
/// Helper function to find a @ref MessageWindow given the window's ID.
/// </summary>
/// <param name="windowId">The ID of the window to find.</param>
/// <returns>Returns a pointer to the @ref MessageWindow that matches the ID;
/// otherwise, returns NULL to indicate no window exists by that ID.</returns>
static MessageWindow* _FindWindow(int windowId)
{
    MessageWindow* foundWindow = NULL;

    if (_windowList != NULL)
    {
        MessageWindow* window = _windowList;
        while (window != NULL)
        {
            if (window->_windowId == windowId)
            {
                foundWindow = window;
                break;
            }
            window = window->next;
        }
    }
    return foundWindow;
}

//-----------------------------------------------------------------------------
//-----------------------------------------------------------------------------

/// <summary>
/// Helper function to handle the ButtonDown message.
/// </summary>
/// <param name="window">The @ref MessageWindow that is handling the message</param>
/// <param name="message">A Message object describing the ButtonDown message.</param>
/// <returns>Always returns false even if the message was handled
/// (allows other windows to select/deselect themselves).</returns>
static bool _HandleButtonDownMessage(MessageWindow* window, Message* message)
{
    // Note: we are not saying we handled the message here since
    // we want other windows to get the button down message as
    // well so they can select or deselect themselves.
    bool messageProcessed = false;

    if (window != NULL && message != NULL)
    {
        if (WindowRectangle_PointInside(&window->_windowBox, &message->Position))
        {
            if (!window->_selected)
            {
                window->_selected = true;
                printf("  --> Button Down in \"%s\", window selected\n", window->_title);
            }
        }
        else
        {
            if (window->_selected)
            {
                window->_selected = false;
                printf("  --> Button Down not in \"%s\", window deselected\n", window->_title);
            }
        }
    }

    return messageProcessed;
}


/// <summary>
/// Helper function to handle the ButtonUp message.
/// </summary>
/// <param name="window">The @ref MessageWindow that is handling the message</param>
/// <param name="message">A Message object describing the ButtonUp message.</param>
/// <returns>Returns true if the message was handled; otherwise, returns
/// false indicating the message was not handled.</returns>
static bool _HandleButtonUpMessage(MessageWindow* window, Message* message)
{
    bool messageProcessed = false;

    if (window != NULL && message != NULL)
    {
        if (window->_selected)
        {
            if (WindowRectangle_PointInside(&window->_windowBox, &message->Position))
            {
                // The Button Up is in the same window as Button Down so
                // we will handle this message and let no other window see
                // it.
                messageProcessed = true;
                if (WindowRectangle_PointInside(&window->_closeBox, &message->Position))
                {
                    printf("  --> Button Up in \"%s\" close box, sending Close message\n", window->_title);
                    Message closeMessage;
                    Message_Initialize(&closeMessage, Close, message->Position.X, message->Position.Y);
                    HandlerChain_SendMessage(window->_windowId, &closeMessage);
                }
                else
                {
                    printf("  --> Button Up in \"%s\", no further action taken\n", window->_title);
                }
            }
        }
    }

    return messageProcessed;
}


/// <summary>
/// Helper function to handle the Close message.
/// </summary>
/// <param name="window">The @ref MessageWindow that is handling the message</param>
/// <param name="message">A Message object describing the Close message.</param>
/// <returns>Returns true if the message was handled; otherwise, returns
/// false indicating the message was not handled.</returns>
static bool _HandleCloseMessage(MessageWindow* window, Message* message)
{
    bool messageProcessed = false;

    if (window != NULL && message != NULL)
    {
        if (window->_selected)
        {
            printf("  --> Close in \"%s\", sending Destroy message\n", window->_title);

            // This window is being closed.  We are handling the message
            // so no other window needs to see it.
            messageProcessed = true;
            Message destroyMessage;
            Message_Initialize(&destroyMessage, Destroy, 0, 0);
            HandlerChain_SendMessage(window->_windowId, &destroyMessage);
        }
        else
        {
            printf("  --> Close seen in \"%s\" but this window is not selected, ignoring\n", window->_title);
        }

    }

    return messageProcessed;
}

/// <summary>
/// Helper function to trigger the destruction of the window.  The window is
/// destroyed and can no longer receive any messages.
///
/// The given @ref MessageWindow is destroyed and can no longer be used after this
/// function returns.
/// </summary>
/// <param name="window">The @ref MessageWindow that is handling the message.</param>
/// <param name="message">The Message object describing the Destroy message.</param>
/// <returns>Returns true if the message was handled; otherwise, returns
/// false indicating the message was not handled.</returns>
static bool _HandleDestroyMessage(MessageWindow* window, Message* message)
{
    bool messageProcessed = false;

    if (window != NULL && message != NULL)
    {
        printf("  --> Destroy in \"%s\", removing window from handler chain and destroying window\n", window->_title);
        HandlerChain_RemoveWindow(window->_windowId);
        window->_selected = false;
        MessageWindow_Destroy(window->_windowId);
        messageProcessed = true;
    }

    return messageProcessed;
}

///////////////////////////////////////////////////////////////////////////////
// MessageWindow_Create()
///////////////////////////////////////////////////////////////////////////////
int MessageWindow_Create(const char* title, int x, int y, int w, int h)
{
    int windowId = -1;

    if (title != NULL)
    {
        MessageWindow* window = calloc(1, sizeof(MessageWindow));
        if (window != NULL)
        {
            windowId = _nextWindowId++;
            window->_windowId = windowId;
            bool success = HandlerChain_AddWindow(window->_windowId);
            if (success)
            {
                window->_title = title;
                WindowRectangle_Initialize(&window->_windowBox, x, y, w, h);
                WindowRectangle_Initialize(&window->_closeBox, window->_windowBox.Right - CLOSE_WIDTH, window->_windowBox.Top, CLOSE_WIDTH, CLOSE_HEIGHT);
                _AppendWindowToList(window);
            }
            else
            {
                printf("  Error!  Out of memory condition adding a window to the handler chain in MessageWindow_Create()!\n");
                free(window);
                window = NULL;
                windowId = -1;
            }
        }
    }

    return windowId;
}


///////////////////////////////////////////////////////////////////////////////
// MessageWindow_Destroy()
///////////////////////////////////////////////////////////////////////////////
void MessageWindow_Destroy(int windowId)
{
    MessageWindow* window = _FindWindow(windowId);
    if (window != NULL)
    {
        _RemoveWindowFromList(window);
        free(window);
    }
}

///////////////////////////////////////////////////////////////////////////////
// MessageWindow_ProcessMessage()
///////////////////////////////////////////////////////////////////////////////
bool MessageWindow_ProcessMessage(int windowId, Message* message)
{
    bool processed = false;

    if (message != NULL)
    {
        MessageWindow* window = _FindWindow(windowId);
        if (window != NULL)
        {
            switch (message->MessageType)
            {
                case Close:
                    processed = _HandleCloseMessage(window, message);
                    break;

                case ButtonDown:
                    processed = _HandleButtonDownMessage(window, message);
                    break;

                case ButtonUp:
                    processed = _HandleButtonUpMessage(window, message);
                    break;

                case Destroy:
                    processed = _HandleDestroyMessage(window, message);
                    break;

                default:
                    printf("Error!  Cannot process unrecognized message type: %d!\n", message->MessageType);
                    processed = true;
                    break;
            }
        }
    }

    return processed;
}


///////////////////////////////////////////////////////////////////////////////
// MessageWindow_ProcessMessage()
///////////////////////////////////////////////////////////////////////////////
bool MessageWindow_ToString(int windowId, DynamicString* output)
{
    bool success = false;

    if (output != NULL)
    {
        MessageWindow* window = _FindWindow(windowId);
        if (window != NULL)
        {
            DynamicString boxOutput = { 0 };
            success = WindowRectangle_ToString(&window->_windowBox, &boxOutput);
            if (success)
            {
                char* buffer = formatstring("[id=%2d] \"%s\" (%s), selected=%s",
                    window->_windowId, window->_title, boxOutput.string,
                    (window->_selected) ? "true" : "false");
                if (buffer != NULL)
                {
                    success = DynamicString_Append(output, buffer);
                    if (!success)
                    {
                        printf("  Error!  Out of memory condition formatting a MessageWindow as a string!\n");
                    }
                    free(buffer);
                }
                else
                {
                    printf("  Error!  Out of memory formatting message window in MessageWindow_ToString()!\n");
                }
            }
            DynamicString_Clear(&boxOutput);
        }
    }

    return success;
}

File: rust/design_pattern_examples_rust/src/handlerchain/handlerchain_messagewindow.rs
//! Contains the WindowRectangle struct and the MessageWindow struct.

use std::cmp::max;
use std::fmt::Display;
use std::sync::atomic::{AtomicI32, Ordering};

use super::handlerchain_message::{MessagePosition, MessageType, Message};
use super::handlerchain_imessagehandler_trait::IMessageHandler;
use super::handlerchain_handlerchain::MessageReturnTypes;

//-----------------------------------------------------------------------------

/// Minimum width of a window (to accommodate a close box).
const MINIMUM_WIDTH: i32 = 4;

/// Minimum height of a window (to accommodate a close box).
const MINIMUM_HEIGHT: i32 = 4;

/// Width of the QUIT region in the upper right corner of the region.
const CLOSE_WIDTH: i32 = 2;
/// Height of the QUIT region in the upper right corner of the region.
const CLOSE_HEIGHT: i32 = 2;

/// Retrieve the next window ID.
fn get_next_window_id() -> i32 {
    // Use a static to hold the next window ID.
    static NEXT_WINDOW_ID: AtomicI32 = AtomicI32::new(1);

    NEXT_WINDOW_ID.fetch_add(1, Ordering::Relaxed)
}

//-----------------------------------------------------------------------------

/// Represents a rectangular region, with upper left and lower right
/// coordinates.
///
/// For this example, the minimum width of a rectangle is 4 x 4.
pub struct WindowRectangle {
        pub left: i32,
        pub top: i32,
        pub right: i32,
        pub bottom: i32,
}

impl WindowRectangle {
    /// Constructor.
    ///
    /// # Parameters
    /// - left
    ///
    ///   X coordinate of upper left corner.
    /// - top
    ///
    ///   Y coordinate of upper left corner.
    /// - width
    ///
    ///   Width of rectangle.
    /// - height
    ///
    ///   Height of rectangle.
    ///
    /// # Returns
    /// Returns a new instance of the WindowRectangle struct.
    pub fn new(left: i32, top: i32, width: i32, height: i32) -> WindowRectangle {
        WindowRectangle {
            left, top,
            right : left + max(width, MINIMUM_WIDTH),
            bottom: top + max(height, MINIMUM_HEIGHT),
        }
    }

    /// Determine if the given point is in the rectangle.
    ///
    /// # Parameters
    /// - point
    ///
    ///   A MessagePosition struct to determine if it is inside this WindowRectangle.
    ///
    /// # Returns
    /// Returns true if the point is in this WindowRectangle; otherwise, returns
    /// false.
    pub fn point_inside(&self, point: &MessagePosition) -> bool {
        let mut is_inside = false;

        if point.x >= self.left && point.x < self.right &&
        point.y >= self.top && point.y < self.bottom {
            is_inside = true;
        }
        is_inside
    }
}

impl Display for WindowRectangle {
    /// Convert this WindowRectangle contents to a string.
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("x1={0:2}, y1={1:2}, x2={2:2}, y2={3:2}", self.left, self.top, self.right, self.bottom))
    }
}

//=============================================================================

/// Represents a rectangular region that can handle messages directed to that
/// region.
///
/// Note: The IMessageHandler interface is an example of a
/// "Facade" design pattern, where the complexity of the MessageWindow is
/// exposed only through a few methods to a system that only needs to deal with
/// those few methods.
pub struct MessageWindow {
    /// Unique ID of this window.
    id: i32,

    /// Title/Name of this window.
    title: String,

    /// Position of this window in global coordinates.
    window_box: WindowRectangle,

    /// Position of the close window within the window box, although the
    /// coordinates are also global coordinates to eliminate the need to
    /// convert between window and global coordinates.
    close_box: WindowRectangle,

    /// Whether this window has been selected (a button click occurred
    /// within the window).
    selected: bool,
}

impl MessageWindow {
    /// Constructor
    ///
    /// # Parameters
    /// - title
    ///
    ///   Title of the MessageWindow.
    /// - x
    ///
    ///   X position of the upper left corner of the window's region.
    /// - y
    ///
    ///   Y position of the upper left corner of the window's region.
    /// - width
    ///
    ///   Width of the window's region.
    /// - height
    ///
    ///   Height of the window's region.
    ///
    /// # Returns
    /// Returns a new instance of the MessageWindow struct.
    pub fn new(title: &str, x: i32, y:i32, width: i32, height: i32) -> MessageWindow {
        MessageWindow {
            id : get_next_window_id(),
            title : title.to_string(),
            window_box: WindowRectangle::new(x, y, width, height),
            close_box: WindowRectangle::new(x + width - CLOSE_WIDTH, y, CLOSE_WIDTH, CLOSE_HEIGHT),
            selected: false,
        }
    }


    /// Helper method to handle the ButtonDown message.
    ///
    /// # Parameters
    /// - message
    ///
    ///   A Message object describing the ButtonDown message.
    ///
    /// # Returns
    /// Returns a value from the MessageReturnTypes enumeration indicating what
    /// action the caller should take.  In this case, always return Continue so
    /// other handlers can react to the same message (assumes no windows are
    /// overlapping).
    fn handle_button_down_message(&mut self, message: &Message) -> MessageReturnTypes {
        if self.window_box.point_inside(&message.message_position) {
            if !self.selected {
                self.selected = true;
                println!("  --> Button Down in \"{0}\", window selected", self.title);
            }
        } else {
            if self.selected {
                self.selected = false;
                println!("  --> Button Down not in \"{0}\", window deselected", self.title);
            }
        }

        // Note: we are saying we didn't handled the message here since
        // we want other windows to get the button down message as
        // well so they can select or deselect themselves.
        MessageReturnTypes::Continue
    }

    /// Helper method to handle the ButtonUp message.
    ///
    /// # Parameters
    /// - message
    ///
    ///   A Message object describing the ButtonUp message.
    ///
    /// # Returns
    /// Returns a value from the MessageReturnTypes enumeration indicating what
    /// action the caller should take: (Stop) This message was handled and
    /// processing should stop, (Continue) this message can be passed on to
    /// other handlers, or (Close) this handler is closed so remove from the
    /// handlers list and stop further processing.
    fn handle_button_up_message(&mut self, message: &Message) -> MessageReturnTypes {
        let mut continue_code = MessageReturnTypes::Continue;

        if self.selected {
            if self.window_box.point_inside(&message.message_position) {
                continue_code = MessageReturnTypes::Stop;
                if self.close_box.point_inside(&message.message_position) {
                    println!("  --> Button Up in \"{0}\" close box, forwarding message to close handler", self.title);
                    continue_code = self.handle_close_message(message);
                } else {
                    println!("  --> Button Up in \"{}\", no further action taken", self.title);
                }
            }
        }
        continue_code
    }

    /// Helper method to handle the Close message.
    ///
    /// # Parameters
    /// - _message
    ///
    ///   NOT USED.
    ///
    /// # Returns
    /// Returns a value from the MessageReturnTypes enumeration indicating what
    /// action the caller should take.  In this case, always return Close to
    /// indicate this handler should be removed from the handler list so no
    /// further messages can be sent to this window.
    fn handle_close_message(&mut self, _message: &Message) -> MessageReturnTypes {
        let mut continue_code = MessageReturnTypes::Continue;

        if self.selected {
            println!("  --> Close in \"{}\", removing window from handler chain", self.title);
            // This window is being closed.  We are handling the message
            // so no other window needs to see it.
            continue_code = MessageReturnTypes::Close;
            self.selected = false;
        }
        continue_code
    }
}


impl IMessageHandler for MessageWindow {
    fn id(&self) -> i32 {
        self.id
    }

    fn process_message(&mut self, message: &super::handlerchain_message::Message) -> MessageReturnTypes {
        match message.message_type {
            MessageType::ButtonDown => self.handle_button_down_message(message),
            MessageType::ButtonUp => self.handle_button_up_message(message),
        }
    }

    fn to_string(&self) -> String {
        format!("[id={:2}] \"{}\" ({}), selected={}", self.id, self.title, self.window_box, self.selected)
    }
}
